Add support for Python slice objects for `std::vector`

cmastalli opened the pull request:
The current  `__getitem__` only accepts integers, which means Eigenpy does not yet support Python slicing. So a slice object (xs[1:]) hits the "Invalid index type" path, as triggered in today's Crocoddyl pre-commit PR (see https://github.com/loco-3d/crocoddyl/actions/runs/17928428896/job/50980225223?pr=1433).

This PR introduces a slice-aware `__getitem__`. The basic strategy is to add an overload that accepts `boost::python::slice` and returns a Python list (with element references if you need mutability, mirroring our single-index path).

I also included a copyright update, but I'm happy to remove it if you guys feel it's not justified.

cmastalli committed:
[vector] Supported Python sliding in std::vector

ManifoldFR commented:
Hi Carlos,
Thanks a lot for this! I'd added slice support for a type in one of my packages some time ago and I never realised we never had this in eigenpy.

I'll take care of reviewing this PR @jcarpent @nim65s.

@cmastalli can you add a test for this feature?

ManifoldFR commented on include/eigenpy/std-vector.hpp:
```diff
@@ -104,6 +106,29 @@ struct overload_base_get_item_for_std_vector
     return bp::object(bp::handle<>(convert(*i)));
   }
 
+  static boost::python::object base_get_item_slice(
```
I think it's also possible to support list indexing (using `bp::list`), i.e.
```python
L = [0, 4, 5]
x = np.arange(1, 10)  # or a list
x[L]  # should be array([1, 5, 6])
```
Would you have time to do that?

cmastalli replied:
In the new version, we are not only handling list indexing but also tuple indexing. See the unit tests!

ManifoldFR replied:
Good, this aligns it with numpy arrays who can be indexed in both ways

(vanilla Python lists don't support list/tuple indexing but my suggestion was to allow it anyway).

ManifoldFR commented on include/eigenpy/std-vector.hpp:
```diff
@@ -104,6 +106,29 @@ struct overload_base_get_item_for_std_vector
+    const Py_ssize_t n = static_cast<Py_ssize_t>(container.get().size());
+
+    if (PySlice_GetIndicesEx(slice.ptr(), n, &start, &stop, &step, &slicelen) != 0)
+      bp::throw_error_already_set();
```
I think you should use the `slice.get_indices(const RandomAccessIterator &begin, const RandomAccessInterator &end)` API instead, see
https://www.boost.org/doc/libs/latest/libs/python/doc/html/reference/object_wrappers/boost_python_slice_hpp.html. It's supposed to handle bound problems.

cmastalli replied:
Done!

cmastalli committed:
[vector] Replace with slice.get_indices

cmastalli committed:
[unittest] Added slicing tests

cmastalli committed:
[vector] Supported list/tuple indexing in vector slicing

cmastalli committed:
[changelog] Updated

cmastalli commented:
@ManifoldFR -- please note that apart from handling your comments, I included a Cmake command to enable us to run the Python tests individually.

jorisv commented on unittest/CMakeLists.txt:
```diff
@@ -73,6 +73,15 @@ function(add_python_eigenpy_lib_unit_test name source)
   set(test_target ${PROJECT_NAME}-${name})
   add_python_unit_test(${test_target} ${source} "lib" "bin" "python")
   set_tests_properties(${test_target} PROPERTIES DEPENDS ${PYWRAP})
+  string(REPLACE "unittest/" "" source_rel "${source}")
+  add_custom_target(
+    "test-${name}"
+    ${CMAKE_COMMAND}
+    -E
+    env
+    PYTHONPATH=${PROJECT_BINARY_DIR}/lib:${PROJECT_BINARY_DIR}/python:$ENV{PYTHONPATH}
+    ${PYTHON_EXECUTABLE}
+    "${CMAKE_CURRENT_SOURCE_DIR}/${source_rel}")
```
```suggestion
```

You can remove this code. `add_python_unit_test` is already setting the appropriate PYTHONPATH to the test.

cmastalli replied:
Thanks for the information! I just reverted it!

ManifoldFR commented on include/eigenpy/std-vector.hpp:
```diff
@@ -104,6 +108,79 @@ struct overload_base_get_item_for_std_vector
+      boost::python::back_reference<Container&> container,
+      boost::python::slice slice) {
+
+    namespace bp = boost::python;
```
I don't think this line is necessary, as there are uses of the `bp::` shorthand before that line and that code worked.

cmastalli replied:
Nice catch! I removed it.

ManifoldFR commented on include/eigenpy/std-vector.hpp:
```diff
@@ -104,6 +108,79 @@ struct overload_base_get_item_for_std_vector
+    typename bp::to_python_indirect<value_type&,
+                                    bp::detail::make_reference_holder> conv;
+    return bp::object(bp::handle<>(conv(c[i])));
+  }
```
Will this work with most types we bind? @jorisv

ManifoldFR commented on CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
+- Supported Python slicing in std::vector in ([#592](https://github.com/stack-of-tasks/eigenpy/pull/592))
```
```suggestion

### Added

- Support for Python slice, tuple and list indexing for `std::vector` bindings ([#592](https://github.com/stack-of-tasks/eigenpy/pull/592))
```

cmastalli replied:
Done!

cmastalli commented:
Hey @ManifoldFR  and @jorisv! I have addressed your comments; however, I am unable to make changes to my commits as I am using a MacBook (i.e., the default filesystem is case-insensitive).

In a nutshell, Git can’t represent the `accelerate.hpp` vs `Accelerate.hpp` files at once and gets "confused." This means that I cannot perform `git rebase -i` as it thinks the files are unstaged. I suggest (1) merging these two files to be compliant with macOS developers in another PR, and (2) squashing my commits when merging.

jorisv commented:
Hello @cmastalli,

You're right about accelerate. We keep the old `accelerate.hpp` header for back compatibility but didn't notice this is an issue for non case sensitive OS.

@ManifoldFR I see three options:
- We move `Accelerate.hpp` to `accelerate.hpp`. This preserve EigenPy API <= 3.11 API.
- We rename `Accelerate.hpp` into `AccelerateSomething.hpp`
- We remove `accelerate.hpp`. This break the API but since I don't think many people are using it, it's maybe not an issue…

ManifoldFR commented:
I say we go with the third option @jorisv. Accelerate isn't even something someone running a packaged version of an eigenpy release would be using - because the package would depend on Eigen 3.4 at most, which doesn't have Accelerate support! So breaking the include is a-ok in my book.

ManifoldFR commented:
Carlos can you fix the [NPY002](https://docs.astral.sh/ruff/rules/numpy-legacy-random/) error the ruff linter reports?

jorisv commented:
@ManifoldFR @cmastalli Accelerate.h has been fixed and I have rebase this PR on devel

cmastalli committed:
[unittest] Fixed NPY002 error

cmastalli commented:
> Carlos can you fix the [NPY002](https://docs.astral.sh/ruff/rules/numpy-legacy-random/) error the ruff linter reports?

Done!

cmastalli commented:
@jorisv and @ManifoldFR -- I have squashed a few commits and fixed the NPY002 error. It could be good to merge now!

ManifoldFR commented:
Ruff and clang-format are still complaining in the pre-commit run. Did you have pre-commit installed in your checked-out repo?

ManifoldFR commented:
Otherwise it will be good to merge

cmastalli committed:
[pre-commit] Run pre-commit

cmastalli commented:
> Otherwise it will be good to merge

Awesome! I didn't know that Eigenpy is using pre-commit. I just pushed a commit that runs `pre-commit` linting.
